crates/noirc_evaluator/src/ssa_refactor/opt/unrolling.rs

```diff
@@ -0,0 +1,575 @@
+            }
+
+            // Don't try to unroll the loop again if it is known to fail
+            if !self.failed_to_unroll.contains(&next_loop.header) {
+                if unroll_loop(function, &self.cfg, &next_loop).is_ok() {
+                    self.modified_blocks.extend(next_loop.blocks);
+                } else {
+                    self.failed_to_unroll.insert(next_loop.header);
+                }
+            }
+        }
+    }
+}
+
+/// Return each block that is in a loop starting in the given header block.
+/// Expects back_edge_start -> header to be the back edge of the loop.
+fn find_blocks_in_loop(
+    header: BasicBlockId,
+    back_edge_start: BasicBlockId,
+    cfg: &ControlFlowGraph,
+) -> Loop {
+    let mut blocks = HashSet::new();
+    blocks.insert(header);
+
+    let mut insert = |block, stack: &mut Vec<BasicBlockId>| {
+        if !blocks.contains(&block) {
+            blocks.insert(block);
+            stack.push(block);
+        }
+    };
+
+    // Starting from the back edge of the loop, each predecessor of this block until
+    // the header is within the loop.
+    let mut stack = vec![];
+    insert(back_edge_start, &mut stack);
+
+    while let Some(block) = stack.pop() {
+        for predecessor in cfg.predecessors(block) {
+            insert(predecessor, &mut stack);
+        }
+    }
+
+    Loop { header, back_edge_start, blocks }
+}
+
+/// Unroll a single loop in the function
+fn unroll_loop(function: &mut Function, cfg: &ControlFlowGraph, loop_: &Loop) -> Result<(), ()> {
+    let mut unroll_into = get_pre_header(cfg, loop_);
+    let mut jump_value = get_induction_variable(function, unroll_into)?;
+
+    while let Some(context) = unroll_loop_header(function, loop_, unroll_into, jump_value) {
+        let (last_block, last_value) = context.unroll_loop_iteration();
+        unroll_into = last_block;
+        jump_value = last_value;
+    }
+
+    Ok(())
+}
+
+/// The loop pre-header is the block that comes before the loop begins. Generally a header block
+/// is expected to have 2 predecessors: the pre-header and the final block of the loop which jumps
+/// back to the beginning.
+fn get_pre_header(cfg: &ControlFlowGraph, loop_: &Loop) -> BasicBlockId {
+    let mut pre_header = cfg
+        .predecessors(loop_.header)
+        .filter(|predecessor| *predecessor != loop_.back_edge_start)
+        .collect::<Vec<_>>();
+
+    assert_eq!(pre_header.len(), 1);
+    pre_header.remove(0)
+}
+
+/// Return the induction value of the current iteration of the loop, from the given block's jmp arguments.
+///
+/// Expects the current block to terminate in `jmp h(N)` where h is the loop header and N is
+/// a Field value.
+fn get_induction_variable(function: &Function, block: BasicBlockId) -> Result<ValueId, ()> {
+    match function.dfg[block].terminator() {
+        Some(TerminatorInstruction::Jmp { arguments, .. }) => {
+            assert_eq!(arguments.len(), 1);
+            let value = arguments[0];
```

Comment:
Can you put an error message in the assert_eq, so that we know why it should never fail.

So:

```
assert_eq!(arguments.len(), 1, "reason why this should not fail")
```

Reply:
Was going to suggest:

```rust
let value = arguments.first().expect("reason why this is always 1")
``` 

But its not exactly the same thing, `first` guarantees that there are atleast one